core/src/main/java/io/confluent/rest/handlers/TenantPrefixSniHandler.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright 2014 - 2023 Confluent Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.confluent.rest.handlers;
+
+import static io.confluent.rest.handlers.SniHandler.getSniServerName;
+import static org.eclipse.jetty.http.HttpStatus.Code.MISDIRECTED_REQUEST;
+
+import java.io.IOException;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import org.eclipse.jetty.server.Request;
+import org.eclipse.jetty.server.handler.HandlerWrapper;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class TenantPrefixSniHandler extends HandlerWrapper {
+
+  private static final Logger log = LoggerFactory.getLogger(TenantPrefixSniHandler.class);
+  private static final String DOT_SEPARATOR = ".";
+  private static final String DASH_SEPARATOR = "-";
+
+  @Override
+  public void handle(String target, Request baseRequest,
+      HttpServletRequest request,
+      HttpServletResponse response) throws IOException, ServletException {
+    String hostHeader = request.getServerName();
+    String sniServerName = getSniServerName(baseRequest);
+
+    if (sniServerName != null) {
+      // Extract the tenantID from the sniServerName, which is always the first segment before '.'
+      // Example: "lsrc-123.us-east-1.aws.private.confluent.cloud" → "lsrc-123"
+      String tenantID = getFirstPart(sniServerName);
+      // The logical cluster ID should also appear at the start of the hostHeader.
+      // It may be followed by either a dot (.) or a dash (-).
+      // hostHeader format examples:
+      // - "lsrc-123-$dom<slug>.us-east-1.aws.glb.confluent.cloud"
+      // - "lsrc-123.$dom<slug>.us-east-1.aws.aws.confluent.cloud"
+      if (tenantID == null
+              || !(hostHeader.startsWith(tenantID + DOT_SEPARATOR)
```

Comment:
can't we just do hostHeader.startsWith(tenantID)?

Reply:
I added separate to disallow,  say `lsrc-123.us-east-1.aws.private.confluent.cloud` to match `lsrc-1234.$dom<slug>.us-east-1.aws.aws.confluent.cloud` in case different Lsrc ID have same prefix